### ISSUE TITLE
Browserify integration is broken

### ISSUE THREAD

**Issue by area** [ipfs/js-ipfs-api#190]
I have been using the example provided at https://github.com/ipfs/js-ipfs-api for a proof-of-concept, successfully communicating with an IPFS node hosted locally for the last week or so. Earlier today, I blew away my `node_modules` directory and reinstalled the relevant packages (i.e. `js-ipfs-api`). The example now no longer works for me - I get an error in my console:

```
Uncaught Error: options.payload must be a string, a Buffer, or a Stream
exports.contain.exports.reachTemplate.exports.assert.condition @ app.js:36055
internals.Client.request.onError @ app.js:47850
requestAPI @ app.js:37627add @ app.js:36985
ColonyJS.store @ app.js:29396
ColonyJS.makeProposal @ app.js:29314
(anonymous function) @ app.js:29265
```

I am aware that due to my use of `browserify`, these line numbers are not useful! Some investigation reveals, I believe, that something fishy is happening around https://github.com/ipfs/js-ipfs-api/blob/285a7e3746110c1fe843b938c026161856c36bcd/src/request-api.js#L83-L85 . `files` is a type `Buffer` using the code, but the `stream` that is returned that I see is of type `Duplexify`. I have not burrowed deep into the dependencies to figure out why this is happening, but it is this conversion that means that the check conducted by Wreck (https://github.com/hapijs/wreck/blob/238647bc9b45120457e35adb85061586b796e0a7/lib/index.js#L71-L73) fails, throwing the error.

I will note with interest that using the wrapper for `js-ipfs-api` at https://github.com/ConsenSys/ipfs.js the example continues to work. I assume this is due to their precise versioning in their `package.json`, and something has inadvertently implemented a breaking change, somewhere.

**Comment by diasdavid** [comment]
What is the IPFS version you are currently running? `ipfs.js` is locked to [`js-ipfs-api@2.3.2`](https://github.com/ConsenSys/ipfs.js/blob/master/package.json#L39), my guess is that you are running a older version of IPFS that has a different HTTP API than the latest release.

**Comment by ckeenan** [comment]
I see this error with the browserified code as well. Tried with builds for 0.3.10, 0.3.11, 0.4.0-dev

`browserify -s ipfsAPI -e ./src/index.js > ipfs.js`

```
<!doctype html>
<html lang="en">
    <body>
        <script src="./ipfs.js"></script>
        <script>
            var ipfs = ipfsAPI('localhost', 5001);
            ipfs.add(new ipfs.Buffer('testing'), function(err, res) {
                console.log('err, res', err, res);
            });
        </script>
    </body>
</html>
```

**Comment by area** [comment]
I'm using IPFS 0.3.10 (installed from `homebrew` on OSX, if that is ultimately relevant)

**Comment by diasdavid** [comment]
@ckeenan do you see it using our build - https://github.com/ipfs/js-ipfs-api/tree/master/dist ?

@area could you try with 0.3.11? https://gobuilder.me/github.com/ipfs/go-ipfs/cmd/ipfs?branch=v0.3.11

**Comment by ckeenan** [comment]
@diasdavid I can [include a dist file in \<script \\\> ](https://github.com/ipfs/js-ipfs-api#in-the-browser-through-script-tag) and that works fine. But trying to [require a browserify-ed version of src](https://github.com/ipfs/js-ipfs-api#in-the-browser-through-browserify) does not

**Comment by diasdavid** [comment]
@ckeenan our dist'ed version suffers some transformations, straight up browserify doesn't work anymore, to check what webpack uses/does, see https://github.com/ipfs/js-ipfs-api/blob/master/tasks/config.js

**Comment by egasimus** [comment]
Having variations of the same issue when using Browserify. Tried passing in a `File`, an `ArrayBuffer`, a [maxogden/filereader-stream](https://github.com/maxogden/filereader-stream), to no avail. Unlikely to have to do anything to do with the server imho, looks totally like a client side thing having to do with Vinyl's type assertions.

**Comment by egasimus** [comment]
Having just gone through a `git bisect`, it turns out that the commit where the browserified library stops recognizing [Browserify buffers](https://github.com/feross/buffer) is, perhaps somewhat unsurprisingly, [the one where Browserify is replaced with Webpack](https://github.com/ipfs/js-ipfs-api/commit/c2a6ad3). Any thoughts on what could've gone wrong @diasdavid, @dignifiedquire?

**Comment by dignifiedquire** [comment]
@egasimus are you using the dist version or are you using a browserified version of `src`?

**Comment by egasimus** [comment]
@dignifiedquire I'm browserifying it myself.

**Comment by diasdavid** [comment]
@dignifiedquire since we had written on the README that browserify is the 'go to' solution to use js-ipfs-api in the browser, we should either 1) remove it (if it is broken completely) and replace by documentation on how to use it with WebPack or 2) even better, make it work with browserify and Webpack and give examples for it (plus tests)

**Comment by jbenet** [comment]
i'd much prefer (2) o/

**Comment by dignifiedquire** [comment]
I talked to @diasdavid and the next action points are:

1. Run all tests through browserify as well as through webpack
2. Setup examples for using js-ipfs-api with webpack and browserify and have them run on CI so we are sure they are up to date and working.

**Comment by dignifiedquire** [comment]
I figured out the issue why this is failing: `vinyl` checks for `isStream` with `instanceof Stream` but this does not work in browserify with a `PassThrough` stream (https://github.com/gulpjs/vinyl/blob/master/lib/isStream.js#L4) I'm not sure how to solve/fix this and we are going to drop vinyl in the 0.4. release anyway which should make this issue go away.

**Comment by egasimus** [comment]
Good to see progress on this. :+1: How do you plan to replace Vinyl?

**Comment by diasdavid** [comment]
We've just released a new version (3.0.1) and also updated the [installation docs](https://github.com/ipfs/js-ipfs-api#installing-the-module), you should not encounter any more problem using this module with browserify. Let us know if it is working for you.

**Comment by diasdavid** [comment]
Closing this one, as it should not be a problem anymore. Let us know if you still have an issue :)
